tracer/src/Datadog.Trace/AppSec/ApiSecurity.cs

```diff
@@ -6,44 +6,101 @@
 #nullable enable
 using System;
 using System.Collections.Generic;
-using System.IO;
-using System.IO.Compression;
-using System.Text;
-using Datadog.Trace.AppSec.Waf;
-using Datadog.Trace.Sampling;
-using Datadog.Trace.Vendors.Newtonsoft.Json;
-using Datadog.Trace.Vendors.Serilog;
+using System.Collections.Specialized;
+using System.Runtime.CompilerServices;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Logging;
 
 namespace Datadog.Trace.AppSec;
 
 internal class ApiSecurity
 {
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<ApiSecurity>();
+    private readonly int _maxRoutesSize;
     private readonly bool _enabled;
-    private readonly OverheadController _overheadController;
+    private readonly TimeSpan _minTimeBetweenReprocessTimeSpan;
+    private readonly OrderedDictionary _processedRoutes = new();
+    private readonly Dictionary<string, bool> _apiSecurityAddress = new() { { "extract-schema", true } };
 
-    public ApiSecurity(SecuritySettings securitySettings)
+    public ApiSecurity(SecuritySettings securitySettings, int maxRouteSize = 4096)
     {
-        _overheadController = new(securitySettings.ApiSecurityMaxConcurrentRequests, (int)(securitySettings.ApiSecuritySampling * 100));
         // todo: later, will be enabled by default, depending on if Security is enabled
         _enabled = securitySettings.ApiSecurityEnabled;
+        _minTimeBetweenReprocessTimeSpan = TimeSpan.FromSeconds(securitySettings.ApiSecuritySampleDelay);
+        _maxRoutesSize = maxRouteSize;
     }
 
-    public bool TryTellWafToAnalyzeSchema(IDictionary<string, object> args)
+    public bool ShouldAnalyzeSchema(bool lastWafCall, Span localRootSpan, IDictionary<string, object> args, string? statusCode, IDictionary<string, object>? routeValues)
     {
-        if (_enabled && _overheadController.AcquireRequest())
+        try
         {
-            args.Add(AddressesConstants.WafContextProcessor, new Dictionary<string, bool> { { "extract-schema", true } });
-            return true;
-        }
+            if (_enabled && lastWafCall && (localRootSpan.Context.TraceContext?.SamplingPriority > (int?)SamplingPriority.AutoReject || localRootSpan.Context.SamplingPriority > (int?)SamplingPriority.AutoReject))
```

Comment:
The `SamplingPriority` enum is deprecated "internally" (meaning we don't want to use it internally anymore, but we don't use `[Obsolete]` yet because it's used in public APIs).

We should use `SamplingPriorityValues` here which avoid the cast to `int` as well.

Reply:
Additionally, we should compare the values for equality, not "greater than". While this works today, it should not be relied on. New values could be added at either end that could break this comparison.

Reply:
I suggest something like this to solve some of my comments above, but this still doesn't fix the issue with a `null` sampling priority due to a delayed decision.

```suggestion
            // localRootSpan.Context.TraceContext should never be null here, but feel free to check to be safe.
            // localRootSpan.Context.TraceContext.SamplingPriority CAN be null here.
            var samplingPriority = localRootSpan.Context.TraceContext.SamplingPriority;

            if (_enabled && lastWafCall && (samplingPriority == SamplingPriorityValues.AutoKeep || samplingPriority == SamplingPriorityValues.UserKeep))
```

Reply:
thanks fixed in [fc9231b](https://github.com/DataDog/dd-trace-dotnet/pull/5257/commits/fc9231bbd11fa8bf58a93c2154bdb7c703b23f2b)